# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: accessible/generic/LocalAccessible.cpp
# Commit: 67f28439cbc7
# Full Hash: 67f28439cbc71117a690ea3eee3ef12b0593a7e6
# Author: Eitan Isaacson <eitan@monotonous.org>
# Date: 2025-09-29 20:54 +0000
# Regressor Bug: 1987945
# File Overlap Count: 1
# Overlapping Files: accessible/tests/browser/relations/browser_anchor_positioning.js
# Description:
#   Bug 1987945 - P3: Support details relation in anchored targets in remote accessibles. r=Jamie,layout-anchor-positioning-reviewers,layout-reviewers,emilio,firefox-style-system-reviewers
# ==============================================================================

diff -r b93b02b24271 -r 67f28439cbc7 accessible/generic/LocalAccessible.cpp
--- a/accessible/generic/LocalAccessible.cpp	Mon Sep 29 20:54:29 2025 +0000
+++ b/accessible/generic/LocalAccessible.cpp	Mon Sep 29 20:54:29 2025 +0000
@@ -1505,9 +1505,15 @@
     mDoc->QueueCacheUpdate(this, CacheDomain::Value);
   }
 
+  if (aAttribute == nsGkAtoms::aria_details) {
+    mDoc->QueueCacheUpdate(this, CacheDomain::Relations);
+    // If an aria-details attribute is added or removed from an anchored
+    // accessible, it will change the validity of its anchor's relation.
+    mDoc->RefreshAnchorRelationCacheForTarget(this);
+  }
+
   if (aAttribute == nsGkAtoms::aria_controls ||
       aAttribute == nsGkAtoms::aria_flowto ||
-      aAttribute == nsGkAtoms::aria_details ||
       aAttribute == nsGkAtoms::aria_errormessage) {
     mDoc->QueueCacheUpdate(this, CacheDomain::Relations);
   }
@@ -4251,6 +4257,11 @@
           if (LocalAccessible* target = GetPopoverTargetDetailsRelation()) {
             rel.AppendTarget(target);
           }
+        } else if (relAtom == nsGkAtoms::target) {
+          if (LocalAccessible* target =
+                  GetAnchorPositionTargetDetailsRelation()) {
+            rel.AppendTarget(target);
+          }
         } else {
           MOZ_ASSERT_UNREACHABLE("Unknown details relAtom");
         }
@@ -4433,6 +4444,22 @@
       mDoc->QueueCacheUpdate(this, CacheDomain::Style);
     }
 
+    if (mOldComputedStyle->StyleDisplay()->mAnchorName !=
+        newStyle->StyleDisplay()->mAnchorName) {
+      // Anchor name changes can affect the result of
+      // details relations.
+      mDoc->QueueCacheUpdate(this, CacheDomain::Relations);
+    }
+
+    if (mOldComputedStyle->MaybeAnchorPosReferencesDiffer(newStyle)) {
+      // Refresh the cache for details on current target (ie. the old style)
+      mDoc->RefreshAnchorRelationCacheForTarget(this);
+      // Refresh the cache for details on new target asynchronously after the
+      // next layout tick for new style.
+      mDoc->Controller()->ScheduleNotification<DocAccessible>(
+          mDoc, &DocAccessible::RefreshAnchorRelationCacheForTarget, this);
+    }
+
     nsAutoCString oldPosition, newPosition;
     mOldComputedStyle->GetComputedPropertyValue(eCSSProperty_position,
                                                 oldPosition);